libsolidity/formal/CHC.cpp

```diff
@@ -226,14 +388,41 @@ smt::SortPointer CHC::interfaceSort()
 	);
 }
 
-smt::SortPointer CHC::functionSort(FunctionDefinition const& _function)
+smt::SortPointer CHC::sort(FunctionDefinition const& _function)
 {
+	if (m_nodeSorts.count(&_function))
+		return m_nodeSorts.at(&_function);
+
 	auto boolSort = make_shared<smt::Sort>(smt::Kind::Bool);
-	auto const& funType = dynamic_cast<FunctionType const&>(*_function.type());
-	return make_shared<smt::FunctionSort>(
-		smt::smtSort(funType.parameterTypes()),
+	vector<smt::SortPointer> varSorts;
+	for (auto const& var: _function.parameters() + _function.returnParameters())
+		varSorts.push_back(smt::smtSort(*var->type()));
+	auto sort = make_shared<smt::FunctionSort>(
+		m_stateSorts + varSorts,
+		boolSort
+	);
+	return m_nodeSorts[&_function] = move(sort);
+}
+
+smt::SortPointer CHC::sort(Block const& _block)
+{
+	if (m_nodeSorts.count(&_block))
+		return m_nodeSorts.at(&_block);
+
+	solAssert(_block.scope() == m_currentFunction, "");
+
+	auto const& fSort = dynamic_pointer_cast<smt::FunctionSort>(sort(*m_currentFunction));
```

Comment:
```suggestion
	auto fSort = dynamic_pointer_cast<smt::FunctionSort>(sort(*m_currentFunction));
```
? Doesn't really matter, though, but the RHS is a temporary anyways... The sort also always lives in ``m_nodeSorts`` as well? Then you could also just use plain pointers... but not too important all that... and can maybe be done in some future refactoring.